fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/search/LuceneOptimizedIndexSearcher.java

```diff
@@ -0,0 +1,115 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.search;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.BulkScorer;
+import org.apache.lucene.search.CollectionTerminatedException;
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.Weight;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+
+/**
+ * This class optimizes the current IndexSearcher and attempts to perform operations in parallel in places where
+ * data access can occur.
+ *
+ *
+ */
+public class LuceneOptimizedIndexSearcher extends IndexSearcher {
+
+    public LuceneOptimizedIndexSearcher(final IndexReader r) {
+        super(r);
+    }
+
+    public LuceneOptimizedIndexSearcher(final IndexReader r, final Executor executor) {
+        super(r, executor);
+    }
+
+    /**
+     * Lower-level search API.
+     *
+     * This is optimized into an attempt to parallelize database access in FDB
+     *
+     * <p>
+     * {@link LeafCollector#collect(int)} is called for every document. <br>
+     *
+     * <p>
+     * NOTE: this method executes the searches on all given leaves exclusively.
+     * To search across all the searchers leaves use {@link #leafContexts}.
+     *
+     * @param leaves
+     *          the searchers leaves to execute the searches on
+     * @param weight
+     *          to match documents
+     * @param collector
+     *          to receive hits
+     * @throws BooleanQuery.TooManyClauses If a query would exceed
+     *         {@link BooleanQuery#getMaxClauseCount()} clauses.
+     */
+    @Override
+    @SuppressWarnings("PMD")
+    protected void search(List<LeafReaderContext> leaves, Weight weight, Collector collector)
+            throws IOException {
+        Executor executor = getExecutor();
+        if (executor == null) {
+            super.search(leaves, weight, collector);
+        } else {
+            // Used to parallelize weight.bulkScorer database access
+            Map<LeafReaderContext, CompletableFuture<BulkScorer>> ctxScorers =
+                    new ConcurrentHashMap<>();
+            leaves.forEach(ctx -> ctxScorers.put(ctx, CompletableFuture.supplyAsync(() -> {
+                try {
+                    collector.getLeafCollector(ctx);
+                    return weight.bulkScorer(ctx);
+                } catch (IOException e) {
+                    return null;
```

Comment:
Do we need to propagate the `IOException` here (based on what the comment below says)?

Reply:
Hmm let me take another look.

Reply:
No, I do not think so since the second pass will trigger at least the pos/freqs exceptions.